src/credentials/DeviceAttestationConstructor.cpp

```diff
@@ -0,0 +1,173 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#include "DeviceAttestationConstructor.h"
+
+#include <lib/core/CHIPTLV.h>
+#include <lib/support/CodeUtils.h>
+#include <lib/support/logging/CHIPLogging.h>
+
+#include <iostream>
+
+namespace chip {
+namespace Credentials {
+
+// context tag positions
+enum
+{
+    CERTIFICATION_DECLARATION = 1,
+    ATTESTATION_NONCE         = 2,
+    TIMESTAMP                 = 3,
+    FIRMWARE_INFO             = 4,
+    NUMBER_OF_VALID_TAGS      = FIRMWARE_INFO,
+};
+
+CHIP_ERROR DeconstructAttestationElements(const ByteSpan & attestationElements, ByteSpan & certificationDeclaration,
+                                          ByteSpan & attestationNonce, uint32_t & timestamp, ByteSpan & firmwareInfo,
+                                          std::vector<ByteSpan> & vendorReserved, uint16_t & vendorId, uint16_t & profileNum)
```

Comment:
std::vector is generally not ok in chip code that will run on embedded due to heap usage.

Reply:
If this was in a separate file pair that was in a build target only used on Commissioners, it would be OK, but right now I agree this could/should be avoided. A good way to avoid it would be to have a vendorReserved data iterator that can be called separately, against the same CSV, to iterate the vendorReserved data.

Reply:
removed the vector and use a simple array.

In the future, we'll use a 
class DAVendorReserved *vendorReserved

replacing

bytespan, &vendorId, &profileNUm